Sources/KavaPlugin+AnalyticsPluginProtocol.swift

```diff
@@ -0,0 +1,218 @@
+// particular library is specified in the applicable library path.
+//
+// You may obtain a copy of the License at
+// https://www.gnu.org/licenses/agpl-3.0.html
+// ===================================================================================================
+
+
+import PlayKit
+
+extension KavaPlugin: AnalyticsPluginProtocol {
+    
+    /************************************************************/
+    // MARK: - AnalyticsPluginProtocol
+    /************************************************************/
+    
+    public var isFirstPlay: Bool {
+        get {
+            return self.kavaData.isFirstPlay
+        }
+        set(newValue) {
+            self.kavaData.isFirstPlay = newValue
+        }
+    }
+    
+    public var playerEventsToRegister: [PlayerEvent.Type] {
+        return [
+            PlayerEvent.stateChanged,
+            PlayerEvent.loadedMetadata,
+            PlayerEvent.play,
+            PlayerEvent.pause,
+            PlayerEvent.playing,
+            PlayerEvent.seeking,
+            PlayerEvent.sourceSelected,
+            PlayerEvent.ended,
+            PlayerEvent.playbackInfo,
+            PlayerEvent.trackChanged,
+            PlayerEvent.error
+        ]
+    }
+    
+    public func registerEvents() {
+        PKLog.debug("register player events")
+        
+        self.playerEventsToRegister.forEach { event in
+            PKLog.debug("Register event: \(event.self)")
+            
+            switch event {
+            case let e where e.self == PlayerEvent.stateChanged:
+                self.messageBus?.addObserver(self, events: [e.self]) { [weak self] event in
+                    guard let strongSelf = self else { return }
+                    PKLog.debug("state changed event: \(event)")
+                    if type(of: event) == PlayerEvent.stateChanged {
+                        strongSelf.handleStateChanged(event: event)
+                    }
+                }
+                
+            case let e where e.self == PlayerEvent.loadedMetadata:
+                self.messageBus?.addObserver(self, events: [e.self], block: { [weak self] (event) in
+                    guard let strongSelf = self else { return }
+                    PKLog.debug("loadedMetadata event: \(event)")
+                    strongSelf.sendMediaLoaded()
+                })
+                
+            case let e where e.self == PlayerEvent.play:
+                self.messageBus?.addObserver(self, events: [e.self], block: { [weak self] (event) in
+                    guard let strongSelf = self else { return }
+                    PKLog.debug("play event: \(event)")
+                    strongSelf.sendAnalyticsEvent(event: KavaEventType.playRequest)
+                    
+                    if strongSelf.sentPlaybackPoints[KavaEventType.playReached100Percent] == true {
+                        strongSelf.sendAnalyticsEvent(event: KavaEventType.replay)
```

Comment:
won't this fire each time we get play event after `playReached100Percent` has been reached?

Reply:
@vadimkononov  let's discuss this, looks like we will never get to this line since on ended we reset the flag to `false` so there is another problem.